#include "MainWindow.h"
#include "ui_MainWindow.h"
#include <QDebug>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QLabel>
#include <QScrollBar>
#include <QTimer>

/**
 * @brief æ„é€ å‡½æ•°
 */
MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , gameEngine_(nullptr)
    , gameTestWidget_(nullptr)
    , gameOutputText_(nullptr)
    , testSwapButton_(nullptr)
    , backToMenuButton_(nullptr)
    , gameView_(nullptr)
    , gameViewWidget_(nullptr)
    , scoreLabel_(nullptr)
{
    ui->setupUi(this);
    setupUi();
    connectSignals();
}

/**
 * @brief ææ„å‡½æ•°
 */
MainWindow::~MainWindow()
{
    if (gameEngine_) {
        delete gameEngine_;
    }
    if (gameTestWidget_) {
        delete gameTestWidget_;
    }
    if (gameViewWidget_) {
        delete gameViewWidget_;
    }
    delete ui;
}

/**
 * @brief åˆå§‹åŒ–UIç»„ä»¶
 */
void MainWindow::setupUi()
{
    // è®¾ç½®çª—å£æ ‡é¢˜å’Œå¤§å°
    setWindowTitle("æ°´æœæ¶ˆæ¶ˆä¹ - Fruit Crush");
    setMinimumSize(800, 600);
    
    qDebug() << "MainWindow initialized";
}

/**
 * @brief è¿æ¥ä¿¡å·å’Œæ§½
 */
void MainWindow::connectSignals()
{
    // è¿æ¥ä¸»èœå•æŒ‰é’®
    connect(ui->casualModeButton, &QPushButton::clicked, this, &MainWindow::startCasualMode);
    connect(ui->competitionModeButton, &QPushButton::clicked, this, &MainWindow::startCompetitionMode);
    connect(ui->leaderboardButton, &QPushButton::clicked, this, &MainWindow::showLeaderboard);
    connect(ui->achievementsButton, &QPushButton::clicked, this, &MainWindow::showAchievements);
}

/**
 * @brief æ˜¾ç¤ºä¸»èœå•
 */
void MainWindow::showMainMenu()
{
    qDebug() << "Show Main Menu";
    // TODO: å®ç°ä¸»èœå•æ˜¾ç¤ºé€»è¾‘
}

/**
 * @brief å¼€å§‹ä¼‘é—²æ¨¡å¼æ¸¸æˆ
 */
void MainWindow::startCasualMode()
{
    qDebug() << "Start Casual Mode - OpenGL Rendering";
    
    // åˆ›å»ºæ¸¸æˆå¼•æ“
    if (!gameEngine_) {
        gameEngine_ = new GameEngine();
    }
    
    // åˆå§‹åŒ–æ¸¸æˆ
    gameEngine_->initializeGame();
    
    // åˆ›å»ºOpenGLæ¸¸æˆè§†å›¾
    if (!gameViewWidget_) {
        createGameViewWidget();
    }
    
    // è®¾ç½®å¼•æ“
    gameView_->setGameEngine(gameEngine_);
    
    // åˆ‡æ¢åˆ°æ¸¸æˆç•Œé¢
    ui->stackedWidget->addWidget(gameViewWidget_);
    ui->stackedWidget->setCurrentWidget(gameViewWidget_);
}

/**
 * @brief å¼€å§‹æ¯”èµ›æ¨¡å¼æ¸¸æˆ
 */
void MainWindow::startCompetitionMode()
{
    qDebug() << "Start Competition Mode";
    // TODO: å®ç°æ¯”èµ›æ¨¡å¼å¯åŠ¨é€»è¾‘
}

/**
 * @brief æ˜¾ç¤ºæ’è¡Œæ¦œ
 */
void MainWindow::showLeaderboard()
{
    qDebug() << "Show Leaderboard";
    // TODO: å®ç°æ’è¡Œæ¦œæ˜¾ç¤ºé€»è¾‘
}

/**
 * @brief æ˜¾ç¤ºæˆå°±é¡µé¢
 */
void MainWindow::showAchievements()
{
    qDebug() << "Show Achievements";
    // TODO: å®ç°æˆå°±é¡µé¢æ˜¾ç¤ºé€»è¾‘
}

/**
 * @brief åˆ›å»ºæ¸¸æˆæµ‹è¯•ç•Œé¢
 */
void MainWindow::createGameTestWidget()
{
    gameTestWidget_ = new QWidget();
    QVBoxLayout* layout = new QVBoxLayout(gameTestWidget_);
    
    // æ¸¸æˆä¿¡æ¯æ ‡ç­¾
    QLabel* titleLabel = new QLabel("ğŸ æ¸¸æˆå¼•æ“æµ‹è¯• ğŸŠ");
    titleLabel->setAlignment(Qt::AlignCenter);
    QFont titleFont = titleLabel->font();
    titleFont.setPointSize(18);
    titleFont.setBold(true);
    titleLabel->setFont(titleFont);
    layout->addWidget(titleLabel);
    
    // æ¸¸æˆè¾“å‡ºæ–‡æœ¬æ¡†ï¼ˆæ˜¾ç¤ºåœ°å›¾å’Œä¿¡æ¯ï¼‰
    gameOutputText_ = new QTextEdit();
    gameOutputText_->setReadOnly(true);
    gameOutputText_->setMinimumHeight(400);
    QFont monoFont("Courier New", 10);
    gameOutputText_->setFont(monoFont);
    layout->addWidget(gameOutputText_);
    
    // æŒ‰é’®å¸ƒå±€
    QHBoxLayout* buttonLayout = new QHBoxLayout();
    
    testSwapButton_ = new QPushButton("æµ‹è¯•äº¤æ¢ (0,0) <-> (0,1)");
    connect(testSwapButton_, &QPushButton::clicked, this, &MainWindow::testSwap);
    buttonLayout->addWidget(testSwapButton_);
    
    backToMenuButton_ = new QPushButton("è¿”å›ä¸»èœå•");
    connect(backToMenuButton_, &QPushButton::clicked, this, &MainWindow::backToMenu);
    buttonLayout->addWidget(backToMenuButton_);
    
    layout->addLayout(buttonLayout);
}

/**
 * @brief æ˜¾ç¤ºæ¸¸æˆåœ°å›¾
 */
void MainWindow::displayGameMap()
{
    if (!gameEngine_ || !gameOutputText_) {
        return;
    }
    
    QString output;
    const auto& map = gameEngine_->getMap();
    
    // æ˜¾ç¤ºæ¸¸æˆçŠ¶æ€
    output += "===========================================\n";
    output += QString("ğŸ® æ¸¸æˆå¼•æ“çŠ¶æ€\n");
    output += "===========================================\n";
    output += QString("ğŸ’¯ å½“å‰åˆ†æ•°: %1\n").arg(gameEngine_->getCurrentScore());
    output += QString("ğŸ”¥ è¿å‡»æ•°: %1\n").arg(gameEngine_->getComboCount());
    output += QString("âœ… æœ‰å¯ç§»åŠ¨: %1\n").arg(gameEngine_->hasValidMoves() ? "æ˜¯" : "å¦");
    output += "\n";
    
    // æ˜¾ç¤ºåœ°å›¾
    output += "ğŸŒ æ¸¸æˆåœ°å›¾ (8x8):\n";
    output += "   ";
    for (int col = 0; col < MAP_SIZE; col++) {
        output += QString(" %1 ").arg(col);
    }
    output += "\n";
    
    for (int row = 0; row < MAP_SIZE; row++) {
        output += QString(" %1 ").arg(row);
        for (int col = 0; col < MAP_SIZE; col++) {
            const Fruit& fruit = map[row][col];
            QString fruitSymbol;
            
            // æ ¹æ®æ°´æœç±»å‹æ˜¾ç¤ºç¬¦å·
            switch (fruit.type) {
                case FruitType::APPLE:      fruitSymbol = "ğŸ"; break;
                case FruitType::ORANGE:     fruitSymbol = "ğŸŠ"; break;
                case FruitType::GRAPE:      fruitSymbol = "ğŸ‡"; break;
                case FruitType::BANANA:     fruitSymbol = "ğŸŒ"; break;
                case FruitType::WATERMELON: fruitSymbol = "ğŸ‰"; break;
                case FruitType::STRAWBERRY: fruitSymbol = "ğŸ“"; break;
                case FruitType::EMPTY:      fruitSymbol = "â¬œ"; break;
            }
            
            // å¦‚æœæœ‰ç‰¹æ®Šå±æ€§ï¼Œæ·»åŠ æ ‡è®°
            if (fruit.special != SpecialType::NONE) {
                switch (fruit.special) {
                    case SpecialType::LINE_H:  fruitSymbol += "H"; break;
                    case SpecialType::LINE_V:  fruitSymbol += "V"; break;
                    case SpecialType::DIAMOND: fruitSymbol += "D"; break;
                    case SpecialType::RAINBOW: fruitSymbol += "R"; break;
                    default: break;
                }
            }
            
            output += fruitSymbol + " ";
        }
        output += "\n";
    }
    
    output += "\n";
    output += "ğŸ’¡ æç¤º: ç‚¹å‡» 'æµ‹è¯•äº¤æ¢' æŒ‰é’®æ¥æµ‹è¯•æ¸¸æˆå¼•æ“\n";
    output += "===========================================\n";
    
    gameOutputText_->setPlainText(output);
    
    // æ»šåŠ¨åˆ°é¡¶éƒ¨
    gameOutputText_->verticalScrollBar()->setValue(0);
}

/**
 * @brief æµ‹è¯•äº¤æ¢æ“ä½œ
 */
void MainWindow::testSwap()
{
    if (!gameEngine_) {
        return;
    }
    
    qDebug() << "Testing swap (0,0) <-> (0,1)";
    
    // å°è¯•äº¤æ¢ (0,0) å’Œ (0,1)
    bool success = gameEngine_->swapFruits(0, 0, 0, 1);
    
    QString message;
    if (success) {
        message = QString("âœ… äº¤æ¢æˆåŠŸï¼å½“å‰åˆ†æ•°: %1, è¿å‡»: %2")
                 .arg(gameEngine_->getCurrentScore())
                 .arg(gameEngine_->getComboCount());
    } else {
        message = "âŒ äº¤æ¢å¤±è´¥ï¼è¯¥äº¤æ¢ä¸ä¼šäº§ç”ŸåŒ¹é…";
    }
    
    qDebug() << message;
    
    // é‡æ–°æ˜¾ç¤ºåœ°å›¾
    displayGameMap();
    
    // åœ¨è¾“å‡ºæ¡†åº•éƒ¨æ˜¾ç¤ºæ¶ˆæ¯
    gameOutputText_->append("\n" + message);
}

/**
 * @brief è¿”å›ä¸»èœå•
 */
void MainWindow::backToMenu()
{
    qDebug() << "Back to main menu";
    
    // åˆ‡æ¢åˆ°ä¸»èœå•é¡µé¢
    ui->stackedWidget->setCurrentIndex(0);
}

/**
 * @brief åˆ›å»ºOpenGLæ¸¸æˆè§†å›¾
 */
void MainWindow::createGameViewWidget()
{
    gameViewWidget_ = new QWidget();
    QVBoxLayout* layout = new QVBoxLayout(gameViewWidget_);
    layout->setContentsMargins(0, 0, 0, 0);
    
    // åˆ›å»ºOpenGLæ¸¸æˆè§†å›¾
    gameView_ = new GameView(gameViewWidget_);
    layout->addWidget(gameView_);
    
    // æ·»åŠ åº•éƒ¨æ§åˆ¶æ 
    QHBoxLayout* controlLayout = new QHBoxLayout();
    
    // åˆ†æ•°æ˜¾ç¤º
    scoreLabel_ = new QLabel("ğŸ’¯ åˆ†æ•°: 0 | ğŸ”¥ è¿å‡»: 0");
    scoreLabel_->setStyleSheet("QLabel { font-size: 16px; font-weight: bold; color: #FFD700; padding: 10px; }");
    controlLayout->addWidget(scoreLabel_);
    
    controlLayout->addStretch();
    
    // è¿”å›æŒ‰é’®
    QPushButton* backButton = new QPushButton("è¿”å›ä¸»èœå•");
    backButton->setMinimumSize(120, 40);
    connect(backButton, &QPushButton::clicked, this, &MainWindow::backToMenu);
    controlLayout->addWidget(backButton);
    
    layout->addLayout(controlLayout);
    
    // åˆ›å»ºå®šæ—¶å™¨æ›´æ–°åˆ†æ•°
    QTimer* scoreTimer = new QTimer(this);
    connect(scoreTimer, &QTimer::timeout, this, [this]() {
        if (gameEngine_ && scoreLabel_) {
            int score = gameEngine_->getCurrentScore();
            int combo = gameEngine_->getComboCount();
            scoreLabel_->setText(QString("ğŸ’¯ åˆ†æ•°: %1 | ğŸ”¥ è¿å‡»: %2").arg(score).arg(combo));
        }
    });
    scoreTimer->start(100);  // æ¯100msæ›´æ–°ä¸€æ¬¡
    
    qDebug() << "GameView widget created";
}
